### PR TITLE
Support Fibre Channel over Ethernet resource of HPE OneView

### PR DESCRIPTION
**Pull request by fgbulsoni**
##### SUMMARY
Added new oneview_fcoe_network module for managing [HPE OneView Fibre Channel over Ethernet](http://h17007.www1.hpe.com/docs/enterprise/servers/oneview3.0/cic-api/en/api-docs/current/index.html#rest/fcoe-networks) resource and unit tests.

Related issue for more information: [HPE OneView support](https://github.com/ansible/ansible/issues/28354)

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`oneview_fcoe_network`

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 655667527d) last updated 2017/08/17 17:56:40 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vagrant/dev/core_ansible_related/ansible/lib/ansible
  executable location = /home/vagrant/dev/core_ansible_related/ansible/bin/ansible
  python version = 2.7.9 (default, Aug  1 2017, 20:38:26) [GCC 4.8.4]

```

##### ADDITIONAL INFORMATION
Example of usage:
```
- name: Ensure that FCoE Network is present using the default configuration
  oneview_fcoe_network:
    config: "{{ config_file_path }}"
    state: present
    data:
      name: 'Test FCoE Network'
      vlanId: '201'

- name: Update the FCOE network scopes
  oneview_fcoe_network:
    config: "{{ config_file_path }}"
    state: present
    data:
      name: 'New FCoE Network'
      scopeUris:
        - '/rest/scopes/00SC123456'
        - '/rest/scopes/01SC123456'
  delegate_to: localhost

- name: Ensure that FCoE Network is absent
  oneview_fcoe_network:
    config: "{{ config_file_path }}"
    state: absent
    data:
      name: 'New FCoE Network'
```

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure in unit tests (in addition to the sanity test failures): https://app.shippable.com/github/ansible/ansible/runs/33925/2/tests

**Comment by mattclay** [comment]
CI failure due to the shebang in the unit test. Remove the shebang to resolve.

**Comment by dagwieers** [comment]
So I haven't reviewed the module, but I did notice these OneView modules are going into **remote_management/hpe**, rather than **remote_management/oneview**. Since HPE is the vendor, I don't think it's particularly smart to put everything there.

Much like **remote_management/hpilo** is specific to HP iLO
https://github.com/ansible/ansible/tree/devel/lib/ansible/modules/remote_management

So HPE makes other stuff that's not remote_management, e.g. network equipment, clustering software, storage systems. I would simply do what we have always been doing, put the modules belonging to a specific technology under the name of the technology, not the vendor.

**Comment by fgbulsoni** [comment]
> I would simply do what we have always been doing, put the modules belonging to a specific technology under the name of the technology, not the vendor.

@dagwieers thanks for your input, should I just open a PR renaming the folders from hpe to oneview and doing whatever adjustments are needed?

**Comment by dagwieers** [comment]
I don't think anyone would complain if you do the rename as part of e.g. this PR.

Note that e.g. **cloud/amazon** and **cloud/vmware** are also Vendor-named directories, but these are arguably also products within the same area.  The location is also not having an effect on users at all, but does influence the structure for documentation and could also help with giving people maintainership over modules. If everything is inside one **hpe/** directory, you may have to give maintainership on a per-module basis and you can't give a oneview maintainer access to new modules (because that requires an update to **BOTMETA.yml**).

**Comment by fgbulsoni** [comment]
> @fgbulsoni this PR contains more than one new module.

@dagwieers I guess the bot complains with more than one module being modified. Should I remove the last commit and make a new PR for the renaming, or leave this as is?

**Comment by dagwieers** [comment]
Oh shoot. No, you can ignore the bot, I just did :-)

**Comment by fgbulsoni** [comment]
> Oh shoot. No, you can ignore the bot, I just did :-)

@dagwieers okie dokie. Any chance I can get the new module reviewed? It *should* be a pretty straightforward one .:octocat:

File: lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py]: Maybe add `default: present` here.

File: lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py]: I noticed this example runs it from the controller, but the other examples don't.
I prefer to put it everywhere, unless you want to clearly show running it differently (from the name or a comment) on purpose.

Up to you, it's not a requirement, just more clear to the reader IMO.

File: lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py]: Also make that default to `present` and remove the `required=True` option

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py]: short_description shouldn't need a trailing dot, as the index does not have trailing dots.

File: lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py]: I personally prefer content in there, leave it up to the user to abstract this into variables. If you do it for them, they have no clue what is expected there. So put the value there, and make a reasonable choice of how your users/customers would be using it. From a files-directory, with a specific directory structure, what naming convention, ...

You have the unique chance to influence good standard practices here, which will help in supporting customers too. Use it :-)

File: lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py]: Are you sure you want this to be string ?

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_fcoe_network.py]: Nice catch, in this specific case it works as both a string or an integer but it is better to leave it as an integer for clarity.

**Comment by fgbulsoni** [comment]
@dagwieers Thank you for the review, I believe the changes requested have been addressed.
Please let me know if there's something else that should be done in this module, otherwise it is good to go for me. :octocat:

**Review by dagwieers — APPROVED** [review]
LGTM

**Pull request merged**
